File: Modules/BuiltInAlgorithms/Sources/Selection/MinMaxHeapSort.swift
import AlgorithmKit
import SortEngineKit

/// A min-max heap (Atkinson et al.) sorted by repeated max-extraction. Unlike a plain binary
/// heap, alternating tree levels enforce opposite invariants — even-depth ("min") levels are
/// `<=` every descendant, odd-depth ("max") levels are `>=` every descendant — so both the
/// minimum (the root) and the maximum (one of the root's two children) are found in O(1), and
/// `downheap` has to look two levels ahead (children *and* grandchildren) to restore whichever
/// invariant applies at the level it started from.
///
/// `isMinLevel` recovers a position's level parity from its 1-based index's bit length (root is
/// index 1 → bit length 1 → min level; that pattern repeats every level going down) — ports
/// ArrayV's `Integer.numberOfLeadingZeros` trick using `Int.leadingZeroBitCount`, which gives the
/// same bit-length parity regardless of the two languages' different native integer widths.
/// `start` is dropped entirely: ArrayV's own `runSort` only ever calls this with `start = 0`, so
/// carrying the field through would just be unused generality.
///
/// Only `store_max` is ported — ArrayV's own `store_min` exists in the source but `runSort` never
/// calls it, so it isn't real behavior to reproduce.
///
/// Stability: `false` — heap extraction reorders equal elements relative to each other.
public struct MinMaxHeapSort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "minmaxheapsort")
  public let metadata = AlgorithmMetadata(
    displayName: "Min-Max Heap Sort",
    category: .selection,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 1114, coefficients: [219835, 258.857, 0.0344196],
      measuredSafeCeiling: nil),
    stable: false,
    timeComplexity: ComplexityBounds(
      best: "O(n log n)", average: "O(n log n)", worst: "O(n log n)"),
    spaceComplexity: "O(1)",
    iconName: "arrow.up.arrow.down.square"
  )
  public init() {}

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n > 1 else { return }

    var end = n

    func isMinLevel(_ position: Int) -> Bool {
      let index = position + 1
      let bitLength = index.bitWidth - index.leadingZeroBitCount
      return bitLength % 2 == 1
    }

    func downheap(_ start: Int) {
      var i = start
      let isGt = !isMinLevel(i)
      func beats(_ a: Int, _ b: Int) -> Bool {
        isGt ? engine.compare(a, b, by: >) : engine.compare(a, b, by: <)
      }

      var left = 2 * i + 1
      while left < end {
        let right = left + 1
        var nexti = left
        for c in [right, 2 * left + 1, 2 * left + 2, 2 * right + 1, 2 * right + 2] {
          if c >= end { break }
          if beats(c, nexti) { nexti = c }
        }
        if nexti <= right {
          if beats(nexti, i) { engine.swap(nexti, i) }
          return
        }
        guard beats(nexti, i) else { return }
        engine.swap(nexti, i)
        let parent = (nexti - 1) / 2
        if beats(parent, nexti) { engine.swap(nexti, parent) }
        i = nexti
        left = 2 * i + 1
      }
    }

    func heapify() {
      var i = (end - 1) / 2
      while i >= 0 {
        downheap(i)
        i -= 1
      }
    }

    func storeMax() {
      guard end > 1 else { return }
      var imax = 1
      if end > 2, engine.compare(imax, imax + 1, by: <) {
        imax += 1
      }
      end -= 1
      engine.swap(imax, end)
      if imax < end {
        downheap(imax)
      }
    }

    heapify()
    for _ in 0..<(n - 1) {
      storeMax()
    }
  }
}

File: App/Resources/AlgorithmDetails/minmaxheapsort/minmaxheapsort.rb
def bit_length(value)
  length = 0
  while value > 0
    value >>= 1
    length += 1
  end
  length
end

def min_level?(index)
  bit_length(index + 1).odd?
end

def better_than?(a, b, min_level)
  min_level ? a < b : a > b
end

def downheap(array, start, size)
  i = start
  loop do
    min_level = min_level?(i)
    left = (2 * i) + 1
    right = (2 * i) + 2
    return if left >= size

    winner = left
    winner = right if right < size && better_than?(array[right], array[winner], min_level)
    base = (4 * i) + 3
    (0...4).each do |offset|
      gc = base + offset
      winner = gc if gc < size && better_than?(array[gc], array[winner], min_level)
    end
    is_grandchild = winner >= base
    extreme = better_than?(array[winner], array[i], min_level)
    unless is_grandchild
      array[i], array[winner] = array[winner], array[i] if extreme
      return
    end
    if extreme
      array[i], array[winner] = array[winner], array[i]
    else
      return
    end
    parent = (winner - 1) / 2
    if min_level
      array[parent], array[winner] = array[winner], array[parent] if array[winner] > array[parent]
    elsif array[winner] < array[parent]
      array[parent], array[winner] = array[winner], array[parent]
    end
    i = winner
  end
end

def heapify(array, length)
  ((length - 1) / 2).downto(0) do |i|
    downheap(array, i, length)
  end
end

def store_max(array, heap_size)
  return heap_size if heap_size <= 1

  imax = 1
  imax = 2 if heap_size > 2 && array[2] > array[1]
  last = heap_size - 1
  array[imax], array[last] = array[last], array[imax]
  new_size = last
  downheap(array, imax, new_size) if imax < new_size
  new_size
end

def sort(arr)
  n = arr.length
  return if n <= 1

  heapify(arr, n)
  heap_size = n
  (n - 1).times do
    heap_size = store_max(arr, heap_size)
  end
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array
